### PR TITLE
fix #299387: support for screenreaders other than NVDA

### PR DESCRIPTION
**Pull request by MarcSabatella**
Resolves: https://musescore.org/en/node/299387

Most of the MuseScore GUI uses standard Qt widgets and thus is as accessible
as one can reasonably expect - especially now that the palettes have moved to QML.
The main issue thing preventing other screenreaders from being viable is the scoreview.
The accessibility implementation made for the screenreader several years ago
works for NVDA, but a number of small details prevent other screenreaders from working.
This change addresses these, allowing JAWS to work relatively normally,
and Windows Narrator to also work with a manual assist (Narrator + Tab shortcut).
These changes don't adversely affect NVDA, and actually improve its response somewhat,
fixing some of the mysterious "split button" texts that were being read.

The changes here are small but collectivey they make a difference,
both in getting JAWS and Narrator to work better out of the box,
but also in making it easier to further script JAWS to provide enhanced accessibility.
I used the Microsoft Accessibility Insights tool to find and debug many of these issues.

- added object names and accessible text for the container objects in the main window
- fixed the rect() returned by the scoreview to use global screen coordinates
- gave the AccessibleScoreView a role() of StaticText rather than NoRole
- allow Description as well as Value as keys to access the accessibility text()
- allow Qt to determine the window() for the AccessibleScoreView
- override isValid() to be sure it always returns true

These changes allow assistive tools to better associate events with objects,
and objects with windows, which is necessary in order for things to work reliably.
Otherwise, there were cases where the assistive tools would be unable
to find the correct object or window.

Most of the changes should be completely non-controversial and harmless at worst.
The change from NoRole to StaticText yielded best results after trying quite a few options.
This was the only meaningful role read by Windows Narrator at all,
and one of the few recognized as valid and by Accessibility Insights.

There are still cases where the screenreader can get confused about the current/active object,
Sometimes it can be useful to press Alt twice (transfer focus to menu bar and back).
Other times using the mouse seems to kickstart things, 
and other times a restart of MuseScore seems the only recourse.
But things seem *much* better with these changes, in my testing.

I know of no way to create automated tests for this.

### CODE REVIEW

File: mscore/scoreaccessibility.cpp [hunk 1]
```diff
@@ -249,22 +270,28 @@ void ScoreAccessibility::updateAccessibilityInfo()
...
-           !mscore->inspector()->isAncestorOf(qApp->focusWidget()) &&
-           !(mscore->searchDialog() && mscore->searchDialog()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getSelectionWindow() && mscore->getSelectionWindow()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getPlayPanel() && mscore->getPlayPanel()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getSynthControl() && mscore->getSynthControl()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getMixer() && mscore->getMixer()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->searchDialog() && mscore->searchDialog()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getDrumrollEditor() && mscore->getDrumrollEditor()->isAncestorOf(qApp->focusWidget())) &&
-           !(mscore->getPianorollEditor() && mscore->getPianorollEditor()->isAncestorOf(qApp->focusWidget()))) {
+      QWidget* focusWidget = qApp->focusWidget();
```
Comment (MarcSabatella) [inline mscore/scoreaccessibility.cpp]: FWIW, this change is a no-op functionally, but it is more efficient, and it aids in further debugging

File: mscore/scoretab.cpp [hunk 2]
```diff
@@ -53,13 +55,17 @@ ScoreTab::ScoreTab(QList<MasterScore*>* sl, QWidget* parent)
       connect(ag, SIGNAL(triggered(QAction*)), this, SIGNAL(actionTriggered(QAction*)));
 
       tab = new QTabBar(this);
+      tab->setObjectName("primary tab");
```
Comment (mattmcclinch) [inline mscore/scoretab.cpp]: I am guessing you probably did not mean to include a space in the object name. I don't know if it matters or not, but no other object name contains any spaces.

Comment (MarcSabatella) [reply mscore/scoretab.cpp]: Thanks, fixed

File: mscore/musescore.cpp [hunk 3]
```diff
@@ -1069,13 +1069,18 @@ MuseScore::MuseScore()
       connect(ag, SIGNAL(triggered(QAction*)), SLOT(cmd(QAction*)));
 
       mainWindow = new QSplitter;
+      mainWindow->setObjectName("mainwindow");
+      mainWindow->setAccessibleName("main window");
```
Comment (Jojo-Schmitz) [inline mscore/musescore.cpp]: are these texts the screen reader reads? If so they should be translatable (and start with upper case)

File: mscore/scoretab.cpp (hunk 2 above)
Comment (Jojo-Schmitz) [reply mscore/scoretab.cpp]: I still see the space though

File: mscore/musescore.cpp (hunk 3 above)
Comment (MarcSabatella) [reply mscore/musescore.cpp]: Good point.  Under normal circumstances the screen won't read most of these; they are more for the benefit of people writing JAWS scripts and as such might be better *not* translated.  But some of them can indeed be read.  I need to learn more about how the scripting works, but hopefully we can get the scripts to work from the object name alone.

**Comment by MarcSabatella** [comment]
I've marked this as WIP, as we probably shouldn't merge it until I have a chance to work out a few more issues (and add translation to the accessible texts).  I'm in touch with some JAWS scripters who will be able to guide me further - one of whom lives about 10 miles from me, so hopefully we'll be able to work together in person.  I'm motivated to get as much figured out before FOSDEM as I can. 
 But if it's at all possible to see this work make it into 3.4, I'll make sure I keep this PR mergeable even if there is additional work that would need to be postponed.

**Comment by MarcSabatella** [comment]
Current status: this all works well for JAWS when we built against 5.9.8, and it does help some with Narrator and apparently VoiceOver, even if it doesn't read automatically by default.  And I've removed the accessibility texts on the object names as it turns out they aren't needed for JAWS scripting.  So it should be mergeable as is.

However, it's worth noting that Qt 5.9.8 really is problematic for NVDA, the palettes are all but unusable due to crashes (with both NVDA and JAWS).  Qt 5.12.5 fixes the crashes, which is great, and makes a lot of other nice improvements in how the accessibilty info is presented to screen readers.  But unfortunately, this PR does *not* get JAWS to read all by itself when using Qt 5.12.5.  Still not sure why, but I will be meeting with a more experienced JAWS user/scripter on Tuesday and may be able to figure some things out.

**Comment by MarcSabatella** [comment]
... and now I'm fiddling with it again, not mergeble any more.  I do hope to have things figured out well enough by mid-week though.

**Comment by MarcSabatella** [comment]
Further progress, I finally have code that enables Orca to work, with the assist of a small script that needs to be installed.  I will see about getting that script included into Orca; they incldue app-specific scripts like this regularly.  I will also investigate using the same approach for JAWS with Qt 5.12, then when all is sorted out I will add appropriate ifdef's to my code as needed.

**Comment by MarcSabatella** [comment]
Apparently I mucked up the commit history while trying to push my changes for orca.  Will try to sort it out.

**Comment by Jojo-Schmitz** [comment]
So this can get closed then?

**Comment by MarcSabatella** [comment]
Yes.  I was going to do that once the other replacement (#5627) built and passed tests successfully, but even though I'm not there yet, I'm certainly closer with that one than this.

**Pull request closed without merging**